src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp

```diff
@@ -873,19 +874,33 @@ HeapWord* ShenandoahHeap::allocate_memory(ShenandoahAllocRequest& req) {
       result = allocate_memory_under_lock(req, in_new_region);
     }
 
+    if (result == nullptr && !req.is_lab_alloc() && get_gc_no_progress_count() > ShenandoahNoProgressThreshold) {
+      // Shenandoah will grind along for quite a while allocating one
+      // object at a time using non-tlab allocations. This will notify
+      // the collector to start a cycle, but will raise an OOME to the
+      // mutator if the last Full GCs have not made progress.
+      control_thread()->handle_alloc_failure(req, false);
+      return nullptr;
+    }
+
     // Allocation failed, block until control thread reacted, then retry allocation.
     //
     // It might happen that one of the threads requesting allocation would unblock
     // way later after GC happened, only to fail the second allocation, because
     // other threads have already depleted the free storage. In this case, a better
     // strategy is to try again, as long as GC makes progress (or until at least
     // one full GC has completed).
-    size_t original_count = shenandoah_policy()->full_gc_count();
-    while (result == nullptr
-        && (_progress_last_gc.is_set() || original_count == shenandoah_policy()->full_gc_count())) {
+    size_t original_count = get_gc_no_progress_count() + 1;
+    while (result == nullptr && original_count > get_gc_no_progress_count()) {
       control_thread()->handle_alloc_failure(req);
       result = allocate_memory_under_lock(req, in_new_region);
     }
+
+    if (log_is_enabled(Debug, gc, alloc)) {
+      ResourceMark rm;
+      log_debug(gc, alloc)("Thread: %s, Result: " PTR_FORMAT ", Shared: %s, Size: " SIZE_FORMAT ", Original: " SIZE_FORMAT ", Latest: " SIZE_FORMAT,
+                           Thread::current()->name(), p2i(result), BOOL_TO_STR(!req.is_lab_alloc()), req.size(), original_count, get_gc_no_progress_count());
```

Comment:
There is `type_string()` that can be used instead of `is_lab_alloc()` here.

Reply:
TIL - thank you.